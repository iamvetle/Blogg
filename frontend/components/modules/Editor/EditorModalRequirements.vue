<template>
	<UModal :value="modelValue" @input="handleInput">
		<UCard>
			<template #header>
				<div class="prose">
					<h3>{{ title }}</h3>
				</div>
			</template>
			<template #default>
				<div class="prose">
					<ul
						:class="charTitle < 3 ? 'text-error' : ''"
						v-html="sanitzedTitle"
					></ul>
					<ul
						:class="charContent < 50 ? 'text-error' : ''"
						v-html="sanitizedContent"
					></ul>
					<ul
						:class="tagsCount > 3 ? 'text-error' : ''"
						v-html="req3Tags"
					></ul>
				</div>
			</template>
		</UCard>
	</UModal>
</template>

<script setup lang="ts">
//@ts-ignore
import DOMPurify from "dompurify";

/**
 * This works.
 *
 * I have to do v-model on this component though "on-top".
 */
const emit = defineEmits(["close", "update:modelValue"]);

const props = withDefaults(
	defineProps<{
		title?: string;
		charTitle: number;
		charContent: number;
		tagsCount: number;
		modelValue: any;
	}>(),
	{
		title: "Requirements not met",
	},
);

/**
 * Handles the input event from the textarea.
 * Emits an 'update:modelValue' event with the new value, allowing the parent component to update its data.
 * @param {Event} event - The input event object.
 */
 const handleInput = (event: any) => {
    emit('update:modelValue', event.target.value);
};

const req1Title = computed(
	() => `
			<li>
			The
			<strong>title</strong> 
			has to be at least 
			<strong>3</strong> 
			characters long.
			</li>
			<ul>
				<li>
					Currently have <strong>${props.charTitle}</strong> characters.
					</li>
				</ul>
		`,
);

const req2Content = computed(
	() => `	<li>
			The 
			<strong>content</strong> has to be at least <strong>50</strong> characters long.
			</li>
			<ul>
				<li>
					Currently have <strong>${props.charContent}</strong> characters.
					</li>
				</ul>

`,
);

const req3Tags = computed(
	() => `	<li>
			The post cannot have more than <strong>three tags</strong>.
			</li>
			<ul>
				<li>Currently have <strong>${props.tagsCount}</strong> characters.
					</li>
			</ul>.
		`,
);
const sanitzedTitle = DOMPurify.sanitize(req1Title.value);
const sanitizedContent = DOMPurify.sanitize(req2Content.value);

const isOpen = ref(true);

/**
 * Makes sure that when the modal is closed by clicking outside the emit that closes the "showModal" is emit
 * I have to have this to make it possible to click outside of the modal and go back to editing.
 */
watchEffect(() => {
	if (isOpen.value === false) {
		emit("close");
		isOpen.value = true;
	}
});

// sanitizes the input/makes it safe (imported it quick - dont know the reliability of it)
</script>

<style scoped></style>
