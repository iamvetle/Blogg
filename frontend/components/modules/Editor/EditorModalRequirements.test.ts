import EditorModalRequirements from './EditorModalRequirements.vue';
import { mount } from '@vue/test-utils';
import { describe, expect, test, beforeEach, afterEach } from 'vitest';

let wrapper: any;

let titleProp = "you must test something"
let descriptionProp = "test description of something"

const factory = () => {
    return mount(EditorModalRequirements, {
        global: {
            plugins: [],
            components: {},
            mocks: {},
            stubs: {},
        },
        props: {
            modelValue:true,
            title:titleProp,
            description:descriptionProp
        },
        slots: {}
    })
};

describe('Testing the requirements modal used for the text editor', () => {

    beforeEach(() => {
    });

    afterEach(() => {
        if (wrapper) {
            wrapper.unmount();
        }
    });
    test('Should exist', () => {
        wrapper = factory()
        expect(wrapper.exists()).toBe(true)
    })
    test('Should have the UModal from nuxt ui', () => {
        wrapper = factory()
        
        console.log(wrapper.html())
        
        const uModal = wrapper.find("UModal")
        expect(uModal.exists()).toBe(true)
    })
    test('Should have the ucard from nuxt ui', () => {
        wrapper = factory()
        
        const uCard = wrapper.find("UCard")
        expect(uCard.exists()).toBe(true)
    })
    test('Should have the titleprop and it be rendered', () => {
        wrapper = factory()

        expect(wrapper.props("title")).toBeTruthy()
    })
    test('Should have the description prop', () => {
        wrapper = factory()

        expect(wrapper.props("description")).toBeTruthy()
        expect(wrapper.html()).toContain(descriptionProp)
        
    })
    
});