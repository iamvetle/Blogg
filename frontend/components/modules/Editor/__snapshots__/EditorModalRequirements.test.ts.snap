// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Testing the requirements modal used for the text editor > Should match snapshot 1`] = `
"<umodal value=\\"true\\">
  <ucard>
    <div class=\\"prose\\">
      <ul class=\\"\\">

        <li>
          The
          <strong>title</strong>
          has to be at least
          <strong>3</strong>
          characters long.
        </li>
        <ul>
          <li>
            Currently have <strong>8</strong> characters.
          </li>
        </ul>
      </ul>
      <ul class=\\"\\">
        <li>
          The
          <strong>content</strong> has to be at least <strong>50</strong> characters long.
        </li>
        <ul>
          <li>
            Currently have <strong>58</strong> characters.
          </li>
        </ul>

      </ul>
      <ul class=\\"text-error\\">
        <li>
          The post cannot have more than <strong>three tags</strong>.
        </li>
        <ul>
          <li>Currently have <strong>4</strong> characters.
          </li>
        </ul>.
      </ul>
    </div>
  </ucard>
</umodal>"
`;
